keps/sig-node/3960-pod-lifecycle-sleep-action/README.md

```diff
@@ -0,0 +1,363 @@
+```go
+type SleepAction struct {
+	// Seconds is the number of seconds to sleep. 
+	Seconds int32
+}
+```
+
+-  Adding a Sleep field to the LifecycleHandler struct, which represents the duration in seconds that the container should sleep before being terminated during the preStop hook.
+```go
+type LifecycleHandler struct {
+	// Sleep pauses further lifecycle progress for a defined time period.
+	Sleep *SleepAction
+}
+```
+
+- When Kubernetes executes the preStop hook with sleep action, it'll simply sleep for a specific seconds.
+```go
+func (hr *handlerRunner) Run(ctx context.Context, containerID kubecontainer.ContainerID, pod *v1.Pod, container *v1.Container, handler *v1.LifecycleHandler) (string, error) {
+    switch {
+    case handler.Exec != nil:...
+    case handler.HTTPGet != nil:...
+    case handler.Sleep != nil:
+        hr.runSleepHandler(ctx, handler.Sleep.Seconds)
+        return "", nil
+    default:...
+    }
+}
+
+func (hr *handlerRunner) runSleepHandler(ctx context.Context, seconds int32) {
+        time.Sleep(time.Duration(seconds) * time.Second)
+}
+```
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+- Test that the runSleepHandler function sleeps for the correct duration when given a valid duration value.
+- Test that the runSleepHandler function returns without error when given a valid duration value.
+- Test that the validation returns an error when given an invalid duration value (e.g., a negative value).
+- Test that the validation returns an error when given duration is longer than the termination graceperiod.
+- Test that the runSleepHandler function returns immediately when given a duration of zero.
+
+##### Integration tests
+N/A
+
+##### e2e tests
+- Basic functionality
+  1. Create a simple pod with a container that runs a long-running process.
+  2. Add a preStop hook to the container configuration, using the new sleepAction with a specified sleep duration (e.g., 5 seconds).
+  3. Delete the pod and observe the time it takes for the container to terminate.
+  4. Verify that the container sleeps for the specified duration before it is terminated.
+
+- Sleep duration boundary testing
+  1. Create a simple pod with a container that runs a long-running process.
+  2. Add a preStop hook to the container configuration, using the new sleepAction with various sleep durations, including:0 seconds (minimum allowed value), values slightly above the minimum allowed value (to test edge cases).
+  3. For each sleep duration, delete the pod and observe the time it takes for the container to terminate.
+  4. Verify that the container sleeps for the specified duration before it is terminated.
+
+- Interaction with termination grace period
```

Comment:
More precisely - you need to verify that the container is terminated after the min(sleep, grace), even when the grace period is shortened.

Create pod with `preStop = sleep(5)`  and termination grace = 10 (passes validation).
Delete the pod with `--grace-period=2`
Verify that it terminates in 2 seconds, not 5

Repeat for --grace-period=0 --force and verify it terminates immediately